node/impl/full/eth.go

```diff
@@ -187,11 +210,127 @@ func (a *EthModule) EthGasPrice(ctx context.Context) (api.EthInt, error) {
 // 	return api.EthHash{}, nil
 // }
 
-// func (a *EthModule) EthEstimateGas(ctx context.Context, tx api.EthTx, blkParam string) (api.EthInt, error) {
-// 	return api.EthInt(0), nil
-// }
-//
-// func (a *EthModule) EthCall(ctx context.Context, tx api.EthTx, blkParam string) (string, error) {
-// 	return "", nil
-// }
-//
+func (a *EthModule) EthEstimateGas(ctx context.Context, tx api.EthCall, blkParam string) (api.EthInt, error) {
+	return api.EthInt(0), nil
+}
+
+func (a *EthModule) EthCall(ctx context.Context, tx api.EthCall, blkParam string) (string, error) {
+	return "", nil
+}
+
+func (a *EthModule) ethBlockFromFilecoinTipSet(ctx context.Context, ts *types.TipSet, fullTxInfo bool) (api.EthBlock, error) {
+	parent, err := a.Chain.LoadTipSet(ctx, ts.Parents())
+	if err != nil {
+		return api.EthBlock{}, err
+	}
+	parentKeyCid, err := parent.Key().Cid()
+	if err != nil {
+		return api.EthBlock{}, err
+	}
+	parentBlkHash, err := api.EthHashFromCid(parentKeyCid)
+	if err != nil {
+		return api.EthBlock{}, err
+	}
+
+	blkMsgs, err := a.Chain.BlockMsgsForTipset(ctx, ts)
+	if err != nil {
+		return api.EthBlock{}, xerrors.Errorf("error loading messages for tipset: %v: %w", ts, err)
+	}
+
+	block := api.NewEthBlock()
+
+	// this seems to be a very expensive way to get gasUsed of the block. may need to find an efficient way to do it
+	gasUsed := int64(0)
+	for _, blkMsg := range blkMsgs {
+		for _, msg := range append(blkMsg.BlsMessages, blkMsg.SecpkMessages...) {
+			msgLookup, err := a.StateAPI.StateSearchMsg(ctx, types.EmptyTSK, msg.Cid(), api.LookbackNoLimit, true)
+			if err != nil {
+				return api.EthBlock{}, nil
+			}
+			gasUsed += msgLookup.Receipt.GasUsed
+
+			if fullTxInfo {
+				tx, err := a.ethTxFromFilecoinMessageLookup(ctx, msgLookup)
+				if err != nil {
+					return api.EthBlock{}, nil
+				}
+				block.Transactions = append(block.Transactions, tx)
+			} else {
+				hash, err := api.EthHashFromCid(msg.Cid())
+				if err != nil {
+					return api.EthBlock{}, err
+				}
+				block.Transactions = append(block.Transactions, hash.String())
+			}
+		}
+	}
```

Comment:
I'm not sure there's any other way of doing it other than loading every single receipt and adding the gas used, like you're doing here. The problem is two-fold: (a) deferred execution due to tipsets, (b) execution tipset does not broadcast the total gas used in the chain data structure (like the Ethereum block does).

Reply:
@arajasek ideas here would be welcome.